src/components/store/store.tsx

```diff
@@ -0,0 +1,223 @@
+  deletingTodoIds: number[],
+  toggleAllTodoCondition: () => void,
+};
+
+const TodosContext = createContext<TodosContextType>({
+  todos: [],
+  hasError: ErrorOption.AddTodoError,
+  resetHasError: () => {},
+  setError: () => {},
+  addTodo: () => {},
+  removeTodo: () => {},
+  recieveTodos: () => {},
+  setFilterSelected: () => {},
+  filterSelected: FilterOption.All,
+  filteredTodos: [],
+  toggleTodoCondition: () => {},
+  resetDeletingTodoIds: () => {},
+  addDeletingTodoIds: () => {},
+  deletingTodoIds: [],
+  toggleAllTodoCondition: () => {},
+  updateTodoTitle: () => {},
+});
+
+export const TodosContextProvider: React.FC<ProviderProps> = ({
+  children,
+}) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [hasError, setHasError] = useState<ErrorOption | null>(null);
+  const [deletingTodoIds, setDeletingTodoIds] = useState<number[]>([]);
+  const [
+    filterSelected,
+    setFilterSelected,
+  ] = useState<FilterOption>(FilterOption.All);
+
+  const recieveTodos = useCallback(
+    (newTodos: Todo[]) => {
+      setTodos(newTodos);
+    },
+    [],
+  );
+
+  const filteredTodos = filterTodos({ todos, filterSelected });
+
+  const resetDeletingTodoIds = () => {
+    setDeletingTodoIds([]);
+  };
+
+  const addDeletingTodoIds = (newTodoIds: number[]) => {
+    setDeletingTodoIds(prevTodoIds => [...prevTodoIds, ...newTodoIds]);
+  };
+
+  const resetHasError = useCallback(
+    () => setHasError(null),
+    [],
+  );
+
+  const setError = useCallback(
+    (errMessage: ErrorOption) => {
+      setHasError(errMessage);
+
+      setTimeout(() => resetHasError(), 3000);
+    },
+    [resetHasError],
+  );
+
+  const updateTodoTitle = (todoId: number, title: string) => {
+    const copyTodos = [...todos];
+
+    const findedTodo = copyTodos.find(todo => todo.id === todoId) || null;
+
+    if (!findedTodo) {
+      return;
+    }
+
+    const indexFindedTodo = copyTodos.indexOf(findedTodo);
+    const updatedTodo = {
+      ...findedTodo,
+      title,
+    };
+
+    copyTodos[indexFindedTodo] = updatedTodo;
+
+    setTodos(copyTodos);
+  };
+
+  const addTodo = (newTodo: Todo) => {
+    setTodos(currentTodos => [...currentTodos, newTodo]);
+  };
+
+  const removeTodo = (todoId: number) => {
+    setTodos(currentTodos => currentTodos.filter(todo => todo.id !== todoId));
+  };
+
+  const toggleAllTodoCondition = async () => {
+    const findedTodos = todos.filter(outerTodo => {
+      const uncompletedTodos = todos.some(innerTodo => !innerTodo.completed);
+
+      if (uncompletedTodos) {
+        return !outerTodo.completed;
+      }
+
+      return outerTodo.completed;
+    });
+
+    if (findedTodos.length === 0) {
+      return;
+    }
+
+    try {
+      const updatePromises = findedTodos.map(async findedTodo => {
+        const updatedTodo = await apiClient.updateTodo(findedTodo.id, {
+          ...findedTodo,
+          completed: !findedTodo.completed,
+        });
+
+        return updatedTodo;
+      });
```

Comment:
```suggestion
      const updatePromises = findedTodos.map(todo => apiClient.updateTodo(todo.id, {
          ...findedTodo,
          completed: !findedTodo.completed,
        }));
```
As you await them with `Promise.all` you can simplify and don't wait inside the loop